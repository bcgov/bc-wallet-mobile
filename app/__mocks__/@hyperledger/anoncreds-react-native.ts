const AnonCredsReactNative = {
  install: jest.fn(),
}

export default AnonCredsReactNative
