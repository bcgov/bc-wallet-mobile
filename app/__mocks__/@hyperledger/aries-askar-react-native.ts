const AriesAskarReactNative = {
  install: jest.fn(),
}

export default AriesAskarReactNative
