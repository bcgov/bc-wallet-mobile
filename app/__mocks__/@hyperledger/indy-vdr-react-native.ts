const IndyVDRReactNative = {
  install: jest.fn(),
}

export default IndyVDRReactNative
