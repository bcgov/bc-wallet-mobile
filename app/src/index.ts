import {
  translationResources,
  ConfigurationContext,
  types,
  Record,
  indyLedgers,
  defaultConfiguration,
} from 'aries-bifold'
import merge from 'lodash.merge'
import { ReducerAction } from 'react'

import bundles from './assets/branding/credential-branding'
import AddCredentialButton from './components/AddCredentialButton'
import AddCredentialSlider from './components/AddCredentialSlider'
import EmptyList from './components/EmptyList'
import HomeContentView from './components/HomeContentView'
import { useNotifications } from './hooks/notifications'
import en from './localization/en'
import { proofRequestTemplates } from './request-templates'
import Developer from './screens/Developer'
import { pages } from './screens/OnboardingPages'
import PersonCredential from './screens/PersonCredential'
import Splash from './screens/Splash'
import Terms from './screens/Terms'
import { BCDispatchAction } from './store'
import { defaultTheme as theme } from './theme'

const localization = merge({}, translationResources, {
  en: { translation: en },
})

const selectedLedgers = indyLedgers.filter((item) => !item.id.startsWith('Indicio'))
const configuration: ConfigurationContext = {
  ...defaultConfiguration,
  pages,
  splash: Splash,
  terms: Terms,
  homeContentView: HomeContentView,
  credentialListHeaderRight: AddCredentialButton,
  credentialListOptions: AddCredentialSlider,
  credentialEmptyList: EmptyList,
  developer: Developer,
  OCABundleResolver: new types.oca.OCABundleResolver(bundles as unknown as types.oca.Bundles),
  record: Record,
  indyLedgers: selectedLedgers,
  settings: [],
  customNotification: {
    component: PersonCredential,
    onCloseAction: (dispatch?: React.Dispatch<ReducerAction<any>>) => {
      if (dispatch) {
        dispatch({
          type: BCDispatchAction.PERSON_CREDENTIAL_OFFER_DISMISSED,
          payload: [{ personCredentialOfferDismissed: true }],
        })
      }
    },
    pageTitle: 'PersonCredential.PageTitle',
    title: 'PersonCredentialNotification.Title',
    description: 'PersonCredentialNotification.Description',
    buttonTitle: 'PersonCredentialNotification.ButtonTitle',
  },
  useCustomNotifications: useNotifications,
  proofRequestTemplates,
  enableTours: true,
}

export default { theme, localization, configuration }
