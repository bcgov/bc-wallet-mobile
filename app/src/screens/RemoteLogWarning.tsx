import { useTheme, testIdWithKey, CheckBoxRow, Button, ButtonType, Link } from '@hyperledger/aries-bifold-core'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Linking, ScrollView, StyleSheet, Text, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'

import ErrorTextBox from '../components/ErrorTextBox'
import FauxHeader from '../components/FauxHeader'

type RemoteLogWarningProps = {
  onBackPressed: () => void
  onEnablePressed: () => void
}

const RemoteLogWarning: React.FC<RemoteLogWarningProps> = ({ onBackPressed, onEnablePressed }) => {
  const [checked, setChecked] = useState(false)
  const { t } = useTranslation()
  const { TextTheme, ColorPallet } = useTheme()

  const onSubmitPressed = () => {
    onEnablePressed()
  }

  const onPressPrivacyPolicyLink = () => {
    Linking.openURL('https://www2.gov.bc.ca/gov/content/governments/government-id/bc-wallet/privacy')
  }

  const style = StyleSheet.create({
    screenContainer: {
      flex: 1,
      marginTop: 10,
      padding: 20,
      justifyContent: 'space-between',
    },
    // No properties needed, just helpful labels
    contentContainer: {},
    controlsContainer: {},
  })

  return (
    <SafeAreaView style={{ flex: 1, backgroundColor: ColorPallet.brand.primary }} edges={['top', 'right', 'left']}>
      <SafeAreaView style={{ flex: 1, backgroundColor: ColorPallet.brand.primaryBackground }} edges={['bottom']}>
        <FauxHeader title={t('RemoteLogging.ScreenTitle')} onBackPressed={onBackPressed} />
        <ScrollView contentContainerStyle={{ flexGrow: 1, backgroundColor: ColorPallet.brand.primaryBackground }}>
          <View style={style.screenContainer}>
            <View style={style.contentContainer}>
              <Text style={[TextTheme.headingTwo, { marginBottom: 10 }]}>{t('RemoteLogging.Heading')}</Text>
              <Text style={[TextTheme.normal, { marginTop: 10, marginBottom: 0 }]}>
                {t('RemoteLogging.CollectionNoticePart1')}
                <Text style={[TextTheme.normal, { fontWeight: 'bold' }]}>
                  {t('RemoteLogging.CollectionNoticeBold')}
                </Text>
                <Text style={TextTheme.normal}>{t('RemoteLogging.CollectionNoticePart2')}</Text>
              </Text>
              <Link
                style={{ marginBottom: 20 }}
                onPress={onPressPrivacyPolicyLink}
                linkText={t('RemoteLogging.CollectionNoticeLink')}
              />
              <ErrorTextBox>{t('RemoteLogging.CollectionNoticeWarning')}</ErrorTextBox>
            </View>
            <View style={style.controlsContainer}>
              <CheckBoxRow
                title={t('RemoteLogging.CheckBoxTitle')}
                accessibilityLabel={t('RemoteLogging.IAgree')}
                testID={testIdWithKey('IAgree')}
                checked={checked}
                onPress={() => setChecked(!checked)}
                reverse
                titleStyle={{ textAlign: 'right' }}
              />
              <View style={{ paddingTop: 10 }}>
                <Button
                  title={t('RemoteLogging.ButtonTitle')}
                  accessibilityLabel={t('RemoteLogging.ButtonTitle')}
                  testID={testIdWithKey('TurnOn')}
                  disabled={!checked}
                  onPress={onSubmitPressed}
                  buttonType={ButtonType.Primary}
                />
              </View>
            </View>
          </View>
        </ScrollView>
      </SafeAreaView>
    </SafeAreaView>
  )
}

export default RemoteLogWarning
