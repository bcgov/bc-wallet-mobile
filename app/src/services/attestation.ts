import { AnonCredsCredentialOffer } from '@credo-ts/anoncreds'
import {
  Agent,
  BaseEvent,
  CredentialState,
  CredentialEventTypes,
  CredentialExchangeRecord,
  ProofState,
  ProofEventTypes,
  ProofExchangeRecord,
  BaseLogger,
  ConnectionRecord,
} from '@credo-ts/core'
import { BifoldError, BifoldAgent } from '@hyperledger/aries-bifold-core'
import {
  generateKey,
  appleAttestation,
  googleAttestation,
  isPlayIntegrityAvailable,
} from '@hyperledger/aries-react-native-attestation'
import { DeviceEventEmitter, Platform } from 'react-native'
import { getVersion, getBuildNumber, getSystemName, getSystemVersion } from 'react-native-device-info'
import { Subscription } from 'rxjs'

import { removeExistingInvitationIfRequired } from '../helpers/BCIDHelper'
import { credentialsMatchForProof } from '../helpers/credentials'
import { AttestationRequestParams, AttestationResult, requestNonceDrpc, requestAttestationDrpc } from '../helpers/drpc'

const defaultResponseTimeoutInMs = 10000 // DRPC response timeout

export type AttestationMonitorOptions = {
  attestationInviteUrl: string
  attestationCredDefIds: string[]
}

export const AttestationEventTypes = {
  Started: 'AttestationEvent.Started',
  Completed: 'AttestationEvent.Completed',
  FailedHandleOffer: 'AttestationEvent.FailedHandleOffer',
  FailedHandleProof: 'AttestationEvent.FailedHandleProof',
  FailedRequestCredential: 'AttestationEvent.FailedRequestCredential',
} as const

export interface AttestationCredentialFormat {
  attributes: {
    attestationInfo: []
  }
}

interface IndyRequest {
  indy: {
    requested_attributes?: {
      attestationInfo?: {
        names: string[]
        restrictions: { cred_def_id: string }[]
      }
    }
  }
}

interface AnonCredsRequest {
  anoncreds: {
    requested_attributes?: {
      attestationInfo?: {
        names: string[]
        restrictions: { cred_def_id: string }[]
      }
    }
  }
}

interface AttestationProofRequestFormat {
  request: IndyRequest & AnonCredsRequest
}

const AttestationErrorCodes = {
  BadInvitation: 2027,
  ReceiveInvitationError: 2028,
  GeneralProofError: 2029,
  FailedToConnectToAttestationAgent: 2030,
  FailedToFetchNonceForAttestation: 2031,
  FailedToGenerateAttestation: 2032,
  FailedToRequestAttestation: 2033,
  FailedToValidateAttestation: 2034,
} as const

export const isProofRequestingAttestation = async (
  proof: ProofExchangeRecord,
  agent: BifoldAgent,
  attestationCredDefIds: string[]
): Promise<boolean> => {
  const format = (await agent.proofs.getFormatData(proof.id)) as unknown as AttestationProofRequestFormat
  const formatToUse = format.request?.anoncreds ? 'anoncreds' : 'indy'

  return !!format.request?.[formatToUse]?.requested_attributes?.attestationInfo?.restrictions?.some((rstr) =>
    attestationCredDefIds ? attestationCredDefIds.includes(rstr.cred_def_id) : false
  )
}

export class AttestationMonitor {
  private proofSubscription?: Subscription
  private offerSubscription?: Subscription
  private agent: Agent
  private options: AttestationMonitorOptions
  private log?: BaseLogger
  private _attestationWorkflowInProgress = false

  // take in options, agent, and logger. Options should include the attestation service URL
  // and the proof to watch for along with the cred_ef_id of the attestation credentials.
  public constructor(agent: Agent, logger: BaseLogger, options: AttestationMonitorOptions) {
    this.agent = agent
    this.log = logger
    this.options = options
  }

  public get attestationWorkflowInProgress() {
    return this._attestationWorkflowInProgress
  }

  public async start(): Promise<void> {
    this.proofSubscription = this.agent.events
      .observable(ProofEventTypes.ProofStateChanged)
      .subscribe(this.handleProofStateChanged)

    this.offerSubscription = this.agent.events
      .observable(CredentialEventTypes.CredentialStateChanged)
      .subscribe(this.handleCredentialStateChanged)
  }

  public stop(): void {
    this.proofSubscription?.unsubscribe()
    this.offerSubscription?.unsubscribe()
  }

  public requestAttestationCredential = async () => {
    this.log?.info('Fetching attestation credential')

    this._attestationWorkflowInProgress = true
    DeviceEventEmitter.emit(AttestationEventTypes.Started)

    try {
      const connection = await this.connectToAttestationAgent()
      if (!connection) {
        throw new BifoldError(
          'Attestation Service',
          'Unable to connect to the attestation service.',
          'No details provided.',
          AttestationErrorCodes.FailedToConnectToAttestationAgent
        )
      }

      const nonce = await this.fetchNonceForAttestation(connection)
      if (!nonce) {
        throw new BifoldError(
          'Attestation Service',
          'There was a problem with the attestation service.',
          'No details provided.',
          AttestationErrorCodes.FailedToFetchNonceForAttestation
        )
      }

      const attestationObj = await this.generateAttestation(nonce)
      if (!attestationObj) {
        throw new BifoldError(
          'Attestation Service',
          'There was a problem with the attestation service.',
          'No details provided.',
          AttestationErrorCodes.FailedToGenerateAttestation
        )
      }

      const result = await this.requestAttestation(connection, attestationObj)
      if (result.status !== 'success') {
        throw new BifoldError(
          'Attestation Service',
          'There was a problem with the attestation service.',
          'No details provided.',
          AttestationErrorCodes.FailedToValidateAttestation
        )
      }
    } catch (error) {
      this._attestationWorkflowInProgress = false
      this.log?.error('Failed to fetch attestation credential', error as Error)

      DeviceEventEmitter.emit(AttestationEventTypes.FailedRequestCredential, error)
    }
  }

  private handleCredentialStateChanged = async (event: BaseEvent) => {
    const { credentialRecord } = event.payload
    const credential = credentialRecord as CredentialExchangeRecord

    this.log?.info('Handling credential offer')

    try {
      const { offer } = await this.agent.credentials.getFormatData(credential.id)
      const offerData = (offer?.anoncreds ?? offer?.indy) as AnonCredsCredentialOffer

      // do nothing if not an attestation credential
      const { attestationCredDefIds } = this.options
      if (!attestationCredDefIds.includes(offerData?.cred_def_id ?? '')) {
        return
      }

      // if it's a new offer, automatically accept
      if (credential.state === CredentialState.OfferReceived) {
        this.log?.info('Accepting credential offer')
        await this.agent.credentials.acceptOffer({
          credentialRecordId: credential.id,
        })
      }

      // only finish loading state once credential is fully accepted
      if (credential.state === CredentialState.Done) {
        // TODO: credential.offer in flight completed
        this.log?.info('Credential accepted')

        this._attestationWorkflowInProgress = false

        DeviceEventEmitter.emit(AttestationEventTypes.Completed)
      }
    } catch (error) {
      this.log?.error('Failed to handle credential offer', error as Error)

      DeviceEventEmitter.emit(AttestationEventTypes.FailedHandleOffer, error)
    }
  }

  private handleProofStateChanged = async (event: BaseEvent) => {
    const { proofRecord } = event.payload
    const proof = proofRecord as ProofExchangeRecord

    this.log?.info('Handling proof received')

    if (proof.state !== ProofState.RequestReceived) {
      return
    }

    this.log?.info('Checking if proof is requesting attestation')

    try {
      // 1. Is the proof requesting an attestation credential
      if (!(await isProofRequestingAttestation(proof, this.agent, this.options.attestationCredDefIds))) {
        return
      }

      this.log?.info('Proof is requesting attestation')

      // 2. Does the wallet have a valid attestation credential that will
      // satisfy the proof request?
      const required = await this.attestationCredentialRequired(this.agent, proof.id)

      // 3. If yes, do nothing
      if (!required) {
        this.log?.info('Valid credentials already exist, nothing to do')

        return
      }

      // 4. If no, get a new attestation credential
      await this.requestAttestationCredential()
    } catch (error) {
      this.log?.error('Failed to handle proof', error as Error)

      DeviceEventEmitter.emit(AttestationEventTypes.FailedHandleProof, error)
    }
  }

  private async connectToAttestationAgent(): Promise<ConnectionRecord | undefined> {
    const invite = await this.agent.oob.parseInvitation(this.options.attestationInviteUrl)

    if (!invite) {
      this.log?.error('Unable to parse attestation agent invitation')

      throw new BifoldError(
        'Attestation Service',
        'Unable to parse the attestation agent invitation',
        'No details provided.',
        AttestationErrorCodes.BadInvitation
      )
    }

    this.log?.info('Removing existing invitation if required')
    await removeExistingInvitationIfRequired(this.agent, invite.id)

    this.log?.info('Receiving invitation')
    const { connectionRecord } = await this.agent.oob.receiveInvitation(invite)
    if (!connectionRecord) {
      throw new BifoldError(
        'Attestation Service',
        'Unable to accept attestation agent invitation',
        'No details provided.',
        AttestationErrorCodes.BadInvitation
      )
    }

    // this step will fail if there is more than one active connection record between a given wallet and
    // the traction instance which is why we need to `removeExistingInvitationIfRequired` above
    return await this.agent.connections.returnWhenIsConnected(connectionRecord.id)
  }

  private async fetchNonceForAttestation(connection: ConnectionRecord): Promise<string> {
    this.log?.info('Requesting nonce from controller')

    const requestNonceCb = await requestNonceDrpc(this.agent, connection)
    const nonceResponse = await requestNonceCb(defaultResponseTimeoutInMs)

    if (!nonceResponse) {
      this.log?.error('Failed to fetch nonce for attestation, code = none, reason = timeout occurred.')

      throw new BifoldError(
        'Attestation Service',
        'There was a problem with the remote attestation service.',
        'Timeout occurred.',
        AttestationErrorCodes.FailedToFetchNonceForAttestation
      )
    }

    this.log?.info('DRPC nonce response received')

    if (nonceResponse.error) {
      this.log?.error(
        `Failed to fetch nonce for attestation, code = ${nonceResponse.error.code}, reason = ${nonceResponse.error.message}`
      )

      throw new BifoldError(
        'Attestation Service',
        'There was a problem with the remote attestation service.',
        nonceResponse.error.message ?? 'No details provided.',
        AttestationErrorCodes.FailedToFetchNonceForAttestation
      )
    }

    const nonce = nonceResponse.result.nonce

    return nonce
  }

  private async requestAttestation(
    connection: ConnectionRecord,
    attestationObj: AttestationRequestParams
  ): Promise<AttestationResult> {
    this.log?.info('Requesting attestation credential from controller')

    const requestAttestationCb = await requestAttestationDrpc(this.agent, connection, attestationObj)
    const attestationResponse = await requestAttestationCb(defaultResponseTimeoutInMs)

    if (!attestationResponse) {
      this.log?.error('Failed to request attestation, code = none, reason = timeout occurred.')

      throw new BifoldError(
        'Attestation Service',
        'There was a problem with the remote attestation service.',
        'Timeout occurred.',
        AttestationErrorCodes.FailedToRequestAttestation
      )
    }

    this.log?.info('DRPC attestation response received')

    if (attestationResponse.error) {
      this.log?.error(
        `Failed to request attestation, code = ${attestationResponse.error.code}, reason = ${attestationResponse.error.message}`
      )

      throw new BifoldError(
        'Attestation Service',
        'There was a problem with the remote attestation service.',
        attestationResponse.error.message ?? 'No details provided.',
        AttestationErrorCodes.FailedToRequestAttestation
      )
    }

    return attestationResponse.result
  }

  private commonAttestationMessageComponent() {
    const common: Partial<AttestationRequestParams> = {
      app_version: `${getVersion()}-${getBuildNumber()}`,
      os_version: `${getSystemName()} ${getSystemVersion()}`,
    }

    return common
  }

  private async generateAttestation(nonce: string) {
    switch (Platform.OS) {
      case 'ios':
        return this.generateAppleAttestation(nonce)
      case 'android':
        return this.generateGoogleAttestation(nonce)

      default:
        // TODO(jl): throw unsupported platform error
        break
    }
  }

  private async generateAppleAttestation(nonce: string) {
    const common = this.commonAttestationMessageComponent()
    const shouldCacheKey = false

    this.log?.info('Generating key for Apple')
    const keyId = await generateKey(shouldCacheKey)

    this.log?.info('Using Apple on-device attestation')
    const attestationAsBuffer = await appleAttestation(keyId, nonce)
    const attestationRequest = {
      ...common,
      platform: 'apple',
      key_id: keyId,
      attestation_object: attestationAsBuffer.toString('base64'),
    } as AttestationRequestParams

    this.log?.info('On-device Apple attestation complete')

    return attestationRequest
  }

  private async generateGoogleAttestation(nonce: string) {
    const common = this.commonAttestationMessageComponent()

    this.log?.info('Checking if Play Integrity is available')

    const available = await isPlayIntegrityAvailable()
    if (!available) {
      return null
    }

    this.log?.info('Using Google on-device attestation')

    const tokenString = await googleAttestation(nonce)
    const attestationRequest = {
      ...common,
      platform: 'google',
      attestation_object: tokenString,
    } as AttestationRequestParams

    this.log?.info('On-device Google attestation complete')

    return attestationRequest
  }

  private attestationCredentialRequired = async (agent: BifoldAgent, proofId: string): Promise<boolean> => {
    agent.config.logger.info('Fetching proof by id')
    const proof = await agent?.proofs.getById(proofId)
    agent.config.logger.info('Second check if proof is requesting attestation')

    agent.config.logger.info('Checking if credentials match for proof request')
    const credentials = await credentialsMatchForProof(agent, proof)

    if (!credentials) {
      return true
    }

    // TODO:(jl) Should we be checking the length of the attributes matches some
    // expected length in the proof request?
    const format = (credentials.proofFormats.anoncreds ?? credentials.proofFormats.indy) as AttestationCredentialFormat
    if (format) {
      return format.attributes.attestationInfo.length === 0
    }

    return false
  }
}
