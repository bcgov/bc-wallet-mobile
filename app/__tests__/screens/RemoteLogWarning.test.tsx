import { StoreProvider } from '@hyperledger/aries-bifold-core'
import { render } from '@testing-library/react-native'
import React from 'react'

import RemoteLogWarning from '../../src/screens/RemoteLogWarning'
import { AttestationProvider } from '../../src/services/attestation'
import { initialState, reducer } from '../../src/store'

describe('RemoteLogWarning Screen', () => {
  beforeEach(() => {
    // Silence console.error because it will print a warning about Switch
    // "Warning: dispatchCommand was called with a ref ...".
    jest.spyOn(console, 'error').mockImplementation(() => {
      return null
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  test('screen renders correctly', () => {
    const tree = render(
      <StoreProvider initialState={initialState} reducer={reducer}>
        <AttestationProvider>
          <RemoteLogWarning onEnablePressed={() => {}} sessionId={123456} />
        </AttestationProvider>
      </StoreProvider>
    )

    expect(tree).toMatchSnapshot()
  })
})
