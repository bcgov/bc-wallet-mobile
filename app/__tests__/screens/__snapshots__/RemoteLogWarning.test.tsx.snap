// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteLogWarning Screen screen renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    [
      "top",
      "right",
      "left",
    ]
  }
  style={
    {
      "backgroundColor": "#42803E",
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    edges={
      [
        "bottom",
      ]
    }
    style={
      {
        "backgroundColor": "#000000",
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "stretch",
          "backgroundColor": "#42803E",
          "borderBottomWidth": 0,
          "elevation": 0,
          "flex": 1,
          "flexDirection": "row",
          "minHeight": 44,
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 6,
            "width": 0,
          },
          "shadowOpacity": 0.15,
          "shadowRadius": 6,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "justifyContent": "center",
          }
        }
      >
        <View
          accessibilityLabel="Global.Back"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 44,
              "left": 44,
              "right": 44,
              "top": 44,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="com.ariesbifold:id/BackButton"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "marginLeft": 15,
                "marginRight": 0,
                "minHeight": 26,
                "minWidth": 26,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontSize": 26,
                  },
                  undefined,
                  {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              󰅁
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexShrink": 1,
            "justifyContent": "center",
            "marginHorizontal": 16,
            "maxWidth": "68%",
            "width": "100%",
          }
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 28,
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          RemoteLogging.ScreenTitle
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "justifyContent": "center",
          }
        }
      />
    </View>
    <RCTScrollView
      contentContainerStyle={
        {
          "backgroundColor": "#000000",
          "flexGrow": 1,
        }
      }
    >
      <View>
        <View
          style={
            {
              "flex": 1,
              "justifyContent": "space-between",
              "marginTop": 10,
              "padding": 20,
            }
          }
        >
          <View
            style={{}}
          >
            <Text
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontSize": 32,
                    "fontWeight": "bold",
                  },
                  {
                    "marginBottom": 10,
                  },
                ]
              }
            >
              RemoteLogging.Heading
            </Text>
            <Text
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontSize": 18,
                    "fontWeight": "normal",
                  },
                  {
                    "marginBottom": 0,
                    "marginTop": 10,
                  },
                ]
              }
            >
              RemoteLogging.CollectionNoticePart1
              <Text
                style={
                  [
                    {
                      "color": "#FFFFFF",
                      "fontSize": 18,
                      "fontWeight": "normal",
                    },
                    {
                      "fontWeight": "bold",
                    },
                  ]
                }
              >
                RemoteLogging.CollectionNoticeBold
              </Text>
              <Text
                style={
                  {
                    "color": "#FFFFFF",
                    "fontSize": 18,
                    "fontWeight": "normal",
                  }
                }
              >
                RemoteLogging.CollectionNoticePart2
              </Text>
            </Text>
            <Text
              accessibilityLabel="RemoteLogging.CollectionNoticeLink"
              accessibilityRole="link"
              accessible={true}
              onPress={[Function]}
              style={
                [
                  {
                    "alignSelf": "flex-start",
                    "color": "#FFFFFF",
                    "fontSize": 18,
                    "fontWeight": "normal",
                    "textDecorationLine": "underline",
                  },
                  {
                    "marginBottom": 20,
                  },
                ]
              }
              testID="com.ariesbifold:id/RemoteLogging.CollectionNoticeLink"
            >
              RemoteLogging.CollectionNoticeLink
            </Text>
            <View
              style={
                {
                  "backgroundColor": "#313132",
                  "borderColor": "#D8292F",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "paddingHorizontal": 10,
                  "paddingVertical": 20,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignSelf": "flex-start",
                      "marginRight": 10,
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "#D8292F",
                          "fontSize": 30,
                        },
                        undefined,
                        {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
                <Text
                  style={
                    {
                      "alignSelf": "center",
                      "color": "#FFFFFF",
                      "flexShrink": 1,
                      "fontSize": 18,
                      "fontWeight": "normal",
                    }
                  }
                >
                  RemoteLogging.CollectionNoticeWarning
                </Text>
              </View>
            </View>
          </View>
          <View
            style={{}}
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row-reverse",
                  "margin": 10,
                }
              }
            >
              <View
                accessibilityLabel="RemoteLogging.IAgree"
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                hitSlop={
                  {
                    "bottom": 44,
                    "left": 44,
                    "right": 44,
                    "top": 44,
                  }
                }
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
                testID="com.ariesbifold:id/IAgree"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#42803E",
                        "fontSize": 36,
                      },
                      undefined,
                      {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  
                </Text>
              </View>
              <Text
                style={
                  [
                    {
                      "color": "#FFFFFF",
                      "flexShrink": 1,
                      "fontSize": 18,
                      "fontWeight": "normal",
                      "marginLeft": 0,
                      "marginRight": 10,
                    },
                    {
                      "textAlign": "right",
                    },
                  ]
                }
              >
                RemoteLogging.CheckBoxTitle
              </Text>
            </View>
            <View
              style={
                {
                  "paddingTop": 10,
                }
              }
            >
              <View
                accessibilityLabel="RemoteLogging.ButtonTitle"
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": true,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "backgroundColor": "rgba(53, 130, 63, 0.35)",
                    "borderRadius": 4,
                    "opacity": 1,
                    "padding": 16,
                  }
                }
                testID="com.ariesbifold:id/TurnOn"
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#FFFFFF",
                          "fontSize": 18,
                          "fontWeight": "bold",
                          "textAlign": "center",
                        },
                        {
                          "color": "#FFFFFF",
                          "fontSize": 18,
                          "fontWeight": "bold",
                          "textAlign": "center",
                        },
                        false,
                        false,
                      ]
                    }
                  >
                    RemoteLogging.ButtonTitle
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RNCSafeAreaView>
</RNCSafeAreaView>
`;
